### PR TITLE
Update the way we calculate SU*Hours for a project.

### LINKED ISSUE #17: change how we do the maths when calculating SUs
## Motivation
We would like to make the calculation of the SU more accurate when there are pods with large SU ratio discrepancies. Currently, the calculation normalizes these by summing too early.

## Completion Criteria
Update the calculation to no longer normalize the CPU to RAM SU Ratio discrepancy.

## Description
- [x] Update the calculation
- [x] Create a sample of the [calculation](https://docs.google.com/spreadsheets/d/1f1W0dJ6XNA8-soHS3ZQIqMSQriocm0fJlhvzzaj6efI/edit)
- [x] Create a demo with the [Nov](https://docs.google.com/spreadsheets/d/1ARWh_fVamP2YLel5_83lCCHvIp8wcgaIVcZNRppNnxo/edit?pli=1#gid=0) data
- [x] Add one line function to this issue to represent what is done to 2 pods, essentially show that the SU is calculated first then summed instead of the other way around.
- [x] link PR
- Merge PR

## New Calculation
Project SU HR count = RoundUP(SUM(Pod1 SU hour count + Pod2 SU hr count + ...)))

## Completion dates
Desired - 2023-12-22
Required - 2024-01-03

### PR DESCRIPTION
**Pull request by naved001**
The old method normalizes RAM and CPU ratio descrepancies. It calcualted vCPU hours and memory hours for each pod, summed it up at the project level and *then* calculated SU hours ( max(vcpu_hours, memory_hours/4).

The new method avoids this by "assigning" fractional SUs to a pod, calculating SU hours at the pod level before summing it up at the project level. This accurately captures the SU usage.

Closes #17

### CODE REVIEW

**Comment by naved001** [comment]
closes #17

File: openshift_metrics/tests/test_utils.py [hunk 1]
```diff
@@ -464,21 +464,33 @@ def test_write_metrics_log(self, mock_gna):
...
+                           "namespace2,PI2,456,1970-01-01T00:00:00,1970-01-01T01:00:00,1.0,pod4,0.5,0,No GPU,2.0,CPU,OpenShift CPU,0.5\n")
 
         tmp_file_name = "%s/test-metrics-%s.log" % (tempfile.gettempdir(), time.time())
         utils.write_metrics_by_pod(test_metrics_dict, tmp_file_name)
-        f = open(tmp_file_name, "r")
-        self.assertEqual(f.read(), expected_output)
-        f.close()
+        with open(tmp_file_name, "r") as f:
+            self.assertEqual(f.read(), expected_output)
+
```
Comment (larsks) [inline openshift_metrics/tests/test_utils.py:493]: You could substantially simplify this code by doing something like this:

```suggestion
        with tempfile.NamedTemporaryFile(mode="w+") as tmp:
            utils.write_metrics_by_pod(test_metrics_dict, tmp.name)
            self.assertEqual(tmp.read(), expected_output)
```

In particular, this takes care of removing the temporary file; the existing code leaves a bunch of temporary files in `/tmp`.

File: openshift_metrics/tests/test_utils.py [hunk 2]
```diff
@@ -572,9 +584,8 @@ def test_write_metrics_log(self, mock_gna):
 
         tmp_file_name = "%s/test-metrics-%s.log" % (tempfile.gettempdir(), time.time())
         utils.write_metrics_by_namespace(test_metrics_dict, tmp_file_name, "2023-01")
-        f = open(tmp_file_name, "r")
-        self.assertEqual(f.read(), expected_output)
-        f.close()
+        with open(tmp_file_name, "r") as f:
+            self.assertEqual(f.read(), expected_output)
```
Comment (larsks) [inline openshift_metrics/tests/test_utils.py:589]: ...as above...

File: openshift_metrics/tests/test_utils.py
Comment (larsks) [inline openshift_metrics/tests/test_utils.py:1]: If you replace:

```
    @mock.patch('requests.get')
    def test_query_metric(self, mock_get):
```

With:

```
    @mock.patch('requests.get')
    @mock.patch('time.sleep')
    def test_query_metric(self, mock_sleep, mock_get):
```

Then your tests will run almost 30 times faster.

Comment (larsks) [reply openshift_metrics/tests/test_utils.py:1]: (And do the same for `test_query_metric_exception`)

Comment (larsks) [reply openshift_metrics/tests/test_utils.py:1]: Also, you can replace:

```
import mock
```

With:

```
from unittest import mock
```

And then you can drop `mock` from `test-requirements.txt`.

Comment (naved001) [reply openshift_metrics/tests/test_utils.py:1]: > Then your tests will run almost 30 times faster.

excellent call!

**Comment by naved001** [comment]
@larsks thanks for the feedback, I have addressed your comments with another commit.

**Review by larsks — APPROVED** [review]

**Pull request merged**